src/validation.h

```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+ * most-work chainstate ("Active") as well as chainstates which may be in
+ * background use to validate UTXO snapshots.
+ *
+ * A few quick definitions:
+ *
+ * *IBD chainstate*: a chainstate populated via initial block download.
+ *
+ * *Snapshot chainstate*: a chainstate populated by loading in an
+ *    assumed-valid UTXO snapshot.
+ *
+ * *Active chainstate*: the chainstate containing the current most-work
+ *    chain. Consulted by most parts of the system (net_processing,
+ *    wallet) as a reflection of the current chain and UTXO set.
+ *    This may either be an IBD chainstate or a snapshot chainstate.
+ *
+ * *Background validation chainstate*: an IBD chainstate for which the
+ *    IBD process is happening in the background while use of the
+ *    active chainstate allows the rest of the system to function.
+ *
+ * *Validated chainstate*: a chainstate whose validity is not assumed.
+ *    This could be an IBD chainstate, or a snapshot chainstate for
+ *    which background validation (up to the base of the snapshot)
+ *    has completed.
+ */
+class ChainstateManager
+{
+private:
+    //! The chainstate used under normal operation (i.e. "regular" IBD) or,
+    //! if a snapshot is in use, for background validation. Its contents will
+    //! be freed when background validation of the snapshot has completed.
+    std::unique_ptr<CChainState> m_ibd_chainstate;
+
+    //! A chainstate initialized on the basis of a UTXO snapshot. If this is
+    //! non-null, it is always our active chainstate unless proven invalid.
+    std::unique_ptr<CChainState> m_snapshot_chainstate;
+
+    //! Points to either the ibd or snapshot chainstate; indicates our
+    //! most-work chain.
+    CChainState* m_active_chainstate;
+
+    //! The base blockhash of the active UTXO snapshot (if one is being used)
+    uint256 m_snapshot_blockhash{};
+
+    //! If true, the assumed-valid chainstate has been fully validated
+    //! by the background validation chainstate.
+    bool m_snapshot_validated{false};
+
+    // For access to m_active_chainstate.
+    friend CChain& ChainActive();
+
+public:
+    //! Instantiate a new chainstate and assign it based upon whether it is
+    //! from a snapshot.
+    //!
+    //! @param[in] activate   If true, make this new chainstate the active one.
+    //! @param[in] snapshot_blockhash   If given, signify that this chainstate
+    //!                                 is based on a snapshot.
+    CChainState& InitializeChainstate(
+        bool activate = true, const uint256& snapshot_blockhash = uint256());
+
+    //! Get all chainstates currently being used.
+    std::vector<CChainState*> GetAll();
+
+    //! The most-work chain.
+    CChain& ActiveChain() const;
+    int ActiveHeight() const { return ActiveChain().Height(); }
+    CBlockIndex* ActiveTip() const { return ActiveChain().Tip(); }
+
+    bool IsSnapshotActive() const;
+
+    uint256 SnapshotBlockhash() const { return m_snapshot_blockhash; }
```

Comment:
in commit 6a629cfd0290716:

How and why is this different from `return m_active_chainstate->m_snapshot_blockhash;`?

Reply:
I think you're right here, I'll remove this.